text/0000-guard_clause_flow_typing.md

```diff
@@ -0,0 +1,284 @@
+examples will be return a `Result` type that will have it's own custom `Error` type used internally.  _This
+is a simplified configuration file parsing example._
+
+### Wrong (A)
+```rust
+fn read_config1() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+  if let Ok(f) = file {
+    let mut buf_reader = BufReader::new(f);
+    let mut contents = String::new();
+
+    if buf_reader.read_to_string(&mut contents).is_ok() {
+      let mut data: Vec<u8> = vec![];
+
+      for item in contents.
+          split("\n").
+          map(|s| s.to_string()).
+          filter(|s| !s.is_empty()).
+          collect::<Vec<String>>() {
+
+        let num = item.parse::<u8>();
+
+        if let Ok(conf) = num {
+          data.push(conf);
+        } else {
+          return Err(Error::ConfigParseFail);
+        }
+      }
+
+      Ok( Config { data: data } )
+    } else {
+      Err(Error::ConfigLoadFail)
+    }
+  } else {
+    Err(Error::ConfigLoadFail)
+  }
+}
+```
+### Wrong (B)
+```rust
+fn read_config2() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+  match file {
+    Ok(f) => {
+      let mut buf_reader = BufReader::new(f);
+      let mut contents = String::new();
+
+      match buf_reader.read_to_string(&mut contents) {
+        Ok(_) => {
+          let mut data: Vec<u8> = vec![];
+
+          for item in contents.
+              split("\n").
+              map(|s| s.to_string()).
+              filter(|s| !s.is_empty()).
+              collect::<Vec<String>>() {
+
+            let num = item.parse::<u8>();
+
+            match num {
+              Ok(conf) => data.push(conf),
+              _ => { return Err(Error::ConfigParseFail); },
+            }
+          }
+
+          Ok( Config { data: data } )
+        },
+        _ => { Err(Error::ConfigLoadFail) }
+      }
+    },
+    _ => { Err(Error::ConfigLoadFail) }
+  }
+}
+```
+### Wrong (C)
+```rust
+fn read_config3() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+
+  if let Ok(f) = file {
+    let mut buf_reader = BufReader::new(f);
+    let mut contents = String::new();
+
+    if buf_reader.read_to_string(&mut contents).is_ok() {
+      let mut data: Vec<u8> = vec![];
+
+      for item in contents.
+          split("\n").
+          map(|s| s.to_string()).
+          filter(|s| !s.is_empty()).
+          collect::<Vec<String>>() {
+
+        let num = item.parse::<u8>();
+
+        if let Ok(conf) = num {
+          data.push(conf);
+        } else {
+          return Err(Error::ConfigParseFail);
+        }
+      }
+
+      return Ok( Config { data: data } );
+    }
+  }
+
+  Err(Error::ConfigLoadFail)
+}
+```
+
+And here is the correct usage of a _Guard Clause_ which allows us to avoid deeply nested logic.
+
+### Correct
+```rust
+fn read_config4() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+
+  // Correct use of Guard Clause
```

Comment:
I don't see how

```rust
let File = File::open("program.cfg");

if let Err(_) = file { return Err(Error::ConfigLoadFail };

let Ok(f) = file;
```

is preferable to

```rust
let f = match File::open("program.cfg") {
    Ok(f) => f,
    Err(_) => return Err(Error::ConfigLoadFail),
};
```

Reply:
Or

```rust
let f = File::open("program.cfg").map_err(|_| Error::ConfigLoadFail)?;
```

Reply:
My main point on these is that the level of complexity for new software developers just learning how to program will find our current methodologies much harder to grasp as much more is going on in these code samples.  One of the things that I've read within our community is that Rust wants to become easier for people to learn and not need to have so much extra syntax used for common scenarios.

I like the `map_err` myself, but I'm not going to teach people Rust as a first language if this is our primary methodology.

Reply:
If anything it seems like this change would increase the cognitive overhead of reading code. Interaction with algebraic datatypes is currently "local", while this change would make it nonlocal in that you can perform destructuring that would normally not be allowed as long as the control flow somewhere some arbitrary distance earlier in the function made it clear that some `Result` is actually only an `Ok`.

You're not going to teach Rust as a first language if *what* exactly is the primary methodology? The use of methods defined on Result? The use of closures? The use of `?`? The use of `match`?

Reply:
We, as experienced developers, take for granted what we're familiar and comfortable with.  The complexity may be harder for us to see because of our own bias.  So I will use visual cues to help illustrate complexity for non-Rust developers and what they need to take in.  The idea is that small simple steps are easier to grasp than more compounded into the same space.

![match](https://user-images.githubusercontent.com/319081/32996333-3b9c68d2-cd4f-11e7-8030-5c0dad10e8dd.png)
---
![map_err](https://user-images.githubusercontent.com/319081/32996334-3dfc824c-cd4f-11e7-957b-2a814ef3bb97.png)
---
![guard_clause](https://user-images.githubusercontent.com/319081/32996335-404b6680-cd4f-11e7-9745-6dad05df32ad.png)

People who are coming from other languages will majoritively be comfortable with an if condition and a code block.  Some may also come from pattern matching languages.  These two concepts are shared by the vast majority of languages available to program in and this allows more expedient learning of a language.

As Rust developers we love our language and the power each design we have permits us to use.  It's also generally true that people form their own preferences/biases in how they like to implement their code.  This often leads to people in their programming language being resistant to change because "this is the way it is" and therefore "should be".  What looks good to us though  is a bias built over time… and that's not a bad thing, but I believe we should be aware of it and more keen to listen to other input knowing our own bias.

The colors above indicate distinct orders of concepts occurring and what we have to grok when we look at the code.  The `match` and `map_err` examples have roughly 4 phases occurring within a tightly knit space where as the simple assignment and pattern matching, though it may be a little more verbose in lines, keeps the grouped concepts of what's happening down to about two steps and the principles behind them share more with other languages in my opinion.

For us I'm perfectly fine with using `match`, `map_err`, and `?` because this **we** know well.  It is my hope that we can endevour to broaden the power and ability we have within Rust to make the language more welcoming to others.  And I believe broadening the ability of pattern matching and creating a flow-typing environment will feel more natural to many.

I'm not against the way we do things, I merely want to broaden our horizons.  And I think in the end we'll be better for it.

Reply:
But this assumes that the syntax is the limiting function in learning the language rather than the semantics, which doesn't really seem true to me.

Reply:
> As Rust developers we love our language and the power each design we have permits us to use. It's also generally true that people form their own preferences/biases in how they like to implement their code. This often leads to people in their programming language being resistant to change because "this is the way it is" and therefore "should be". What looks good to us though is a bias built over time… and that's not a bad thing, but I believe we should be aware of it and more keen to listen to other input knowing our own bias.

I agree in general, but I would say that a great deal of this bias (at least for me) comes from the fact that Rust syntax tends to be very elegant, expressive, and ergonomic IMHO; I would like to keep it that way. I still spend a lot of time writing C and Java, and it still surprises me how much easier it can be to write rust.

That's not to say I'm unopen to new syntax -- I just don't think we should introduce new idiomatic syntax purely for teachability unless there is a significant community consensus that existing syntax is confusing (e.g. what sort of happened with `dyn Trait`).

Reply:
> You're not going to teach Rust as a first language if what exactly is the primary methodology? The use of methods defined on Result? The use of closures? The use of ?? The use of match?

For a new developer…Of these three I believe `closures` would be the more difficult of these to teach (not for me to say, for others to understand).  `match` would take some time to get across, especially with assignment beforehand, but pattern matching would be the first thing to teach before delving into `match`.  After pattern matching is taught the `?` wouldn't be too hard to teach.    But having them all in one statement would be a bit much for a beginner.

_It's been both my hobby and profession to teach concepts and programming through blog posts for many years.  Thinking about clarifying things is pretty much what I do in life._